Fix for issue #57268 - ENH: Preserve input start/end type in interval…

VISWESWARAN1998 opened the pull request:
…_range

- [ ] closes #57268 
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

VISWESWARAN1998 committed:
Fix for issue #57268 - ENH: Preserve input start/end type in interval_range

VISWESWARAN1998 committed:
issue #57268 - github actions resolution

VISWESWARAN1998 committed:
Use generated datatype from breaks

VISWESWARAN1998 committed:
Ruff - Pre-commit issue fix

mroeschke commented on pandas/core/indexes/interval.py:
```diff
@@ -1120,9 +1120,19 @@ def interval_range(
+                breaks,
+                name=name,
+                closed=closed,
+                dtype=IntervalDtype(subtype=breaks.dtype, closed=closed),
```
Can you just pass a `dtype` to the `IntervalIndex.from_breaks` below?

VISWESWARAN1998 replied:
Yes, I can do that!

VISWESWARAN1998 replied:
All changes have been implemented and tests are passed. Unit test is added as well!

mroeschke reviewed (changes requested):
Thanks for the PR. Needs a whatsnew note in `v3.0.0.rst` and a unit test.

Also could you fix this for float types too?

VISWESWARAN1998 commented:
> ew note in `v3.0.0.rst` and a unit test.
> 
> Also could you fix this for float types too?

Can you give me an example for float type with expected results? Thank you!

VISWESWARAN1998 committed:
Fix for issue #57268 - floating point support

VISWESWARAN1998 commented:
I have added my implementations for supporting float types. But I see in base.py that pandas doesn't support float16 data types. Should I convert them to float64 or just let pandas thow Exception?

base.py:
```python3
elif dtype == np.float16:
    # float16 not supported (no indexing engine)
    raise NotImplementedError("float16 indexes are not supported")
```

For example:

```python3
interval_index_float = pd.interval_range(start=np.float32(0.0), end=np.float32(0.6), freq=0.5, closed="left")
print(interval_index_float)
```

Gives:

```
IntervalIndex([[0.0, 0.5)], dtype='interval[float32, left]')
```

But:

```python3
interval_index_float = pd.interval_range(start=np.float16(0.0), end=np.float16(0.6), freq=0.5, closed="left")
print(interval_index_float)
```

will raise:

`NotImplementedError: float16 indexes are not supported`

I have added implementation to ignore float16 like this and checked-in my changes. If you want to let it throw an exception let me know!

My implementation:

```python3
if isinstance(start, float | np.float16) or isinstance(
                end, float | np.float16
            ):
    dtype = np.dtype("float64")
else:
    dtype = start.dtype if start.dtype == end.dtype else np.dtype("float64")
breaks = np.arange(start, end + (freq * 0.1), freq, dtype=dtype)
```

Current implementation safely returns float64 if it finds float16, Just like current main release!

VISWESWARAN1998 committed:
int - float dtype compatability

VISWESWARAN1998 committed:
whatsnew documentation update

VISWESWARAN1998 committed:
OS based varaible access

VISWESWARAN1998 committed:
Fixing failed unit test cases

VISWESWARAN1998 committed:
pytest - interval passsed

VISWESWARAN1998 committed:
Python backwards compatability

VISWESWARAN1998 committed:
Pytest

VISWESWARAN1998 committed:
Fixing PyLint and mypy issues

VISWESWARAN1998 reviewed (commented):
Completed as you asked!

mroeschke commented on doc/source/whatsnew/v3.0.0.rst:
```diff
@@ -31,7 +31,7 @@ Other enhancements
 - :func:`read_stata` now returns ``datetime64`` resolutions better matching those natively stored in the stata format (:issue:`55642`)
 - Allow dictionaries to be passed to :meth:`pandas.Series.str.replace` via ``pat`` parameter (:issue:`51748`)
--
+- :func:`interval_range` Preserves start/end data types
```
This should go under the `Interval` section

VISWESWARAN1998 replied:
Done!

mroeschke commented on pandas/tests/indexes/interval/test_interval_range.py:
```diff
@@ -220,6 +220,27 @@ def test_float_subtype(self, start, end, freq):
+    )
+    def test_interval_dtype(self, iteration, start, end):
+        result = interval_range(start=start, end=end).dtype.subtype
+        expected = [
```
Can you put this in the `parametrize`?

VISWESWARAN1998 replied:
You mean put expected in the paramaterize?

VISWESWARAN1998 replied:
Please check the new changes

mroeschke commented on pandas/core/indexes/interval.py:
```diff
@@ -1101,9 +1101,25 @@ def interval_range(
     breaks: np.ndarray | TimedeltaIndex | DatetimeIndex
 
     if is_number(endpoint):
+        dtype: Any = np.dtype("int64")
```
```suggestion
        dtype: np.dtype = np.dtype("int64")
```

mroeschke commented on pandas/core/indexes/interval.py:
```diff
@@ -1101,9 +1101,25 @@ def interval_range(
+                or isinstance(freq, (float, np.float16))
+            ):
+                dtype = np.dtype("float64")
+            elif isinstance(start, int) or isinstance(end, int):
```
Covered by your initial assignment

mroeschke commented on pandas/core/indexes/interval.py:
```diff
@@ -1101,9 +1101,25 @@ def interval_range(
+                dtype = np.dtype("float64")
+            elif isinstance(start, int) or isinstance(end, int):
+                dtype = np.dtype("int64")
+            elif np.issubdtype(start.dtype, np.integer) and np.issubdtype(
```
Can simplify a lot of these cases by checking

```
if isinstance(start, (np.integer, np.floating)) and isinstance(end, (np.integer, np.floating)) and start.dtype == end.dtype:
    dtype = start.dtype
```

VISWESWARAN1998 replied:
Done!

VISWESWARAN1998 committed:
Merge branch 'main' into 57268

VISWESWARAN1998 committed:
dtype specification

VISWESWARAN1998 committed:
Conditional statement simplification

VISWESWARAN1998 committed:
remove redundant code blocks

VISWESWARAN1998 committed:
Changing whatsnew to interval section

VISWESWARAN1998 committed:
Passing expected in parameterize

VISWESWARAN1998 commented:
Changes has been completed as you asked!

mroeschke commented on pandas/core/indexes/interval.py:
```diff
@@ -1101,9 +1101,23 @@ def interval_range(
+                dtype = start.dtype
             # 0.1 ensures we capture end
             breaks = np.arange(start, end + (freq * 0.1), freq)
+            breaks = maybe_downcast_numeric(breaks, dtype)
```
I don't think you need this if you just specify `dtype` in `arange`

mroeschke commented on pandas/core/indexes/interval.py:
```diff
@@ -1101,9 +1101,23 @@ def interval_range(
+                or isinstance(freq, (float, np.float16))
+            ):
+                dtype = np.dtype("float64")
+            if (
```
```suggestion
            elif (
```

mroeschke commented on doc/source/whatsnew/v3.0.0.rst:
```diff
@@ -211,7 +211,7 @@ Strings
 Interval
 ^^^^^^^^
--
+- :func:`interval_range` Preserves start/end data types
```
```suggestion
- Bug in :func:`interval_range` where start and end numeric types were always cast to 64 bit (:issue:`57268`)
```

VISWESWARAN1998 committed:
Update doc/source/whatsnew/v3.0.0.rst

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

VISWESWARAN1998 committed:
Update pandas/core/indexes/interval.py

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

VISWESWARAN1998 commented:
You are most welcome @mroeschke !
